wdl-lint/src/rules/disallowed_declaration_name.rs

```diff
@@ -0,0 +1,259 @@
+//! A lint rule that disallows declaration names with type prefixes or suffixes.
+
+use wdl_ast::AstToken;
+use wdl_ast::Diagnostic;
+use wdl_ast::Diagnostics;
+use wdl_ast::Document;
+use wdl_ast::Span;
+use wdl_ast::SupportedVersion;
+use wdl_ast::SyntaxElement;
+use wdl_ast::SyntaxKind;
+use wdl_ast::VisitReason;
+use wdl_ast::Visitor;
+use wdl_ast::v1::BoundDecl;
+use wdl_ast::v1::Decl;
+use wdl_ast::v1::PrimitiveTypeKind;
+use wdl_ast::v1::UnboundDecl;
+
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// The identifier for the disallowed declaration name rule.
+const ID: &str = "DisallowedDeclarationName";
+
+/// Diagnostic for declaration identifiers with type prefixes or suffixes
+fn decl_identifier_with_type(span: Span, decl_name: &str, type_name: &str) -> Diagnostic {
+    Diagnostic::note(format!("declaration identifier '{}' contains type name '{}'", decl_name, type_name))
+        .with_rule(ID)
+        .with_highlight(span)
+        .with_fix("rename the identifier to not include the type name")
+}
+
+/// A rule that checks for declaration names that contain their type.
+#[derive(Debug, Default)]
+pub struct DisallowedDeclarationNameRule;
+
+impl Visitor for DisallowedDeclarationNameRule {
+    type State = Diagnostics;
+
+    fn document(
+        &mut self,
+        _: &mut Self::State,
+        reason: VisitReason,
+        _: &Document,
+        _: SupportedVersion,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry
+        *self = Default::default();
+    }
+
+    fn bound_decl(&mut self, state: &mut Self::State, reason: VisitReason, decl: &BoundDecl) {
+        if reason == VisitReason::Enter {
+            check_decl_name(state, &Decl::Bound(decl.clone()), &self.exceptable_nodes());
+        }
+    }
+
+    fn unbound_decl(&mut self, state: &mut Self::State, reason: VisitReason, decl: &UnboundDecl) {
+        if reason == VisitReason::Enter {
+            check_decl_name(
+                state,
+                &Decl::Unbound(decl.clone()),
+                &self.exceptable_nodes(),
+            );
+        }
+    }
+}
+
+impl Rule for DisallowedDeclarationNameRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Declaration names should not include their type"
+    }
+
+    fn explanation(&self) -> &'static str {
+        "Declaration names should not include their type as a prefix or suffix. \
+        This makes the code more verbose and often redundant. For example, use \
+        'input_file' instead of 'file_input' or 'myFile'."
```

Comment:
@a-frantz - We should rework this explanation since we also disallow `input_file`.